big_scape/trees/newick_tree.py

```diff
@@ -0,0 +1,221 @@
+from typing import TextIO
+from Bio import Phylo
+from pathlib import Path
+from collections import defaultdict
+from scipy.optimize import linear_sum_assignment
+
+# from other modules
+from big_scape.genbank import BGCRecord
+
+
+def generate_newick_tree(
+    records: list[BGCRecord], exemplar: int, family_members: list[int]
+) -> str:
+    """Generate newick formatted tree for each GCF
+
+    Args:
+        records (list[BGCRecord]): list of records within GCF
+        exemplar (int): index of exemplar to use during alignment
+        family_members (list[int]): list of bgc ids in one family
+
+    Returns:
+        Correctly formatted newick tree
+    """
+    # no need for alignment
+    if len(records) < 3:
+        return f"({','.join([str(bgc_id)+':0.0' for bgc_id in family_members])}):0.01;"
+
+    algn = generate_gcf_alignment(records, exemplar, family_members)
+    with tempfile.TemporaryDirectory() as tmpdir:
+        tmppath = Path(tmpdir)
+        tmp_algn = tmppath / Path("alignment.fasta")
+        tmp_newick = tmppath / Path("alignment.newick")
+        with open(tmp_algn, "w") as tmp_a:
+            tmp_a.write(algn)
+        with open(tmp_newick, "w") as tmp_n:
+            run_fasttree(tmp_algn, tmp_n)
+        tree = process_newick_tree(tmp_newick)
+    return tree
+
+
+def run_fasttree(algn_file: Path, out_file: TextIO):
+    """Generate FastTree newick GCF tree
+
+    Args:
+        algn_file (Path): Path to alignment file
+        out_file (TextIO): Opened output file object
+    """
+    subprocess.run(
+        ["fasttree", "-quiet", "-nopr", algn_file], stdout=out_file, shell=False
+    )
+
+
+def process_newick_tree(tree_file: Path) -> str:
+    """Process newick tree file format
+
+    Args:
+        tree_file (Path): Path to tree file
+    """
+    if not tree_file.exists():
+        logging.error("Failed to create newick tree")
+        raise FileNotFoundError()
+    with open(tree_file, "r") as newick_file:
+        try:
+            tree = Phylo.read(newick_file, "newick")
+        except ValueError as e:
+            logging.warning("Error encountered while reading newick tree: ", str(e))
+            return ""
+        else:
+            try:
+                tree.root_at_midpoint()
+            except UnboundLocalError:
+                # Noticed this could happen if the sequences are exactly
+                # the same and all distances == 0
+                logging.debug("Unable to root at midpoint")
+            return tree.format("newick")
+
+
+def generate_gcf_alignment(
+    records: list[BGCRecord], exemplar: int, family_members: list[int]
+) -> str:
+    """Write protein domain alignment to file
+
+    Args:
+        records (list[BGCRecord]): Records within one GCF to align
+        exemplar (int): Index of exemplar to use during alignment
+        family_members (list[int]): list of bgc ids in one family
+
+    Returns:
+        Written file containing protein domain alignment
+    """
```

Comment:
can you add a sneaky TODO: refactor to this docstring? just to indicate that this is not our code :sunglasses: